### PR TITLE
Use unmodified official ClangFormat configuration as base formatter configuration

### PR DESCRIPTION
**Pull request by per1234**
The Arduino IDE's "**Auto Format**" feature is configured to produce the standard Arduino sketch formatting style by default.

The Arduino IDE editor's default settings are compliant with that style. However, the user may adjust the editor settings. In this case, the Arduino IDE automatically adjusts the "**Auto Format**" configuration to align with the user's preferences.

The formatter configuration is consumed by several other projects in addition to the Arduino IDE. For this reason, the configuration is hosted and maintained in [a centralized location](https://github.com/arduino/tooling-project-assets/tree/main/other/clang-format-configuration), from which it is pulled by all projects that use it.

### Motivation

The adjustment of the Arduino IDE formatter configuration according to the editor settings is integrated into the configuration object itself:

https://github.com/arduino/arduino-ide/blob/75e00c2bae231c8319220ac286319f3bbe49c167/arduino-ide-extension/src/node/clang-formatter.ts#L324-L326

This means that the standardized configuration has to be modified each time it is pulled in to sync from the upstream source.

### Change description

Move the adjustment code out of the base formatter configuration object to allow syncs to be done via a pure copy/paste replacement.

### Other information

Example of a configuration sync: https://github.com/arduino/arduino-ide/pull/1285

### Reviewer checklist

* [ ] PR addresses a single concern.
* [ ] The PR has no duplicates (please search among the [Pull Requests](https://github.com/arduino/arduino-ide/pulls) before creating one)
* [ ] PR title and description are properly filled.
* [ ] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

**Comment by kittaakos** [comment]
> ### Change description
> Move the adjustment code out of the base formatter configuration object to allow syncs to be done via a pure copy/paste replacement.

We could move the default formatter JSON into a dedicated file if it helps the synchronization. Consider the following folder structure:

```
arduino-ide-extension/src/node/clang-formatter.ts
arduino-ide-extension/src/node/default-formatter-config.json (<-- you can pick a better name if you want)
```

and then from code:

```ts
function styleJson({
  TabWidth,
  UseTab,
}: ClangFormatOptions): Record<string, unknown> {
  // Source: https://github.com/arduino/tooling-project-assets/tree/main/other/clang-format-configuration
  const defaultConfig = require('./default-formatter-config.json'); // 1. require the JSON
  return {
    ...defaultConfig,
    TabWidth, // 2. override the default values with the user-defined ones
    UseTab,
  };
}
```

 1. Require the JSON content with [`require`](https://nodejs.org/en/knowledge/getting-started/what-is-require/). To do this, we have to add `"resolveJsonModule": true` to `arduino-ide-extension/tsconfig.json`,
 2. Use the spread operator to merge in the user-defined values. (See the docs [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax#spread_in_object_literals)).
 
I think this will make the config synchronization easier and we do not have to touch code when we update the config. What do you think?

<details><summary>Click here, I created and formatted the JSON so that you can copy-paste it into a 
<code>.json</code> file</summary>
<p>

```json
{
    "AccessModifierOffset": -2,
    "AlignAfterOpenBracket": "Align",
    "AlignArrayOfStructures": "None",
    "AlignConsecutiveAssignments": "None",
    "AlignConsecutiveBitFields": "None",
    "AlignConsecutiveDeclarations": "None",
    "AlignConsecutiveMacros": "None",
    "AlignEscapedNewlines": "DontAlign",
    "AlignOperands": "Align",
    "AlignTrailingComments": true,
    "AllowAllArgumentsOnNextLine": true,
    "AllowAllConstructorInitializersOnNextLine": true,
    "AllowAllParametersOfDeclarationOnNextLine": true,
    "AllowShortBlocksOnASingleLine": "Always",
    "AllowShortCaseLabelsOnASingleLine": true,
    "AllowShortEnumsOnASingleLine": true,
    "AllowShortFunctionsOnASingleLine": "Empty",
    "AllowShortIfStatementsOnASingleLine": "AllIfsAndElse",
    "AllowShortLambdasOnASingleLine": "Empty",
    "AllowShortLoopsOnASingleLine": true,
    "AlwaysBreakAfterDefinitionReturnType": "None",
    "AlwaysBreakAfterReturnType": "None",
    "AlwaysBreakBeforeMultilineStrings": false,
    "AlwaysBreakTemplateDeclarations": "No",
    "AttributeMacros": [
        "__capability"
    ],
    "BasedOnStyle": "LLVM",
    "BinPackArguments": true,
    "BinPackParameters": true,
    "BitFieldColonSpacing": "Both",
    "BraceWrapping": {
        "AfterCaseLabel": false,
        "AfterClass": false,
        "AfterControlStatement": "Never",
        "AfterEnum": false,
        "AfterFunction": false,
        "AfterNamespace": false,
        "AfterObjCDeclaration": false,
        "AfterStruct": false,
        "AfterUnion": false,
        "AfterExternBlock": false,
        "BeforeCatch": false,
        "BeforeElse": false,
        "BeforeLambdaBody": false,
        "BeforeWhile": false,
        "IndentBraces": false,
        "SplitEmptyFunction": true,
        "SplitEmptyRecord": true,
        "SplitEmptyNamespace": true
    },
    "BreakAfterJavaFieldAnnotations": false,
    "BreakBeforeBinaryOperators": "NonAssignment",
    "BreakBeforeBraces": "Attach",
    "BreakBeforeConceptDeclarations": false,
    "BreakBeforeInheritanceComma": false,
    "BreakBeforeTernaryOperators": true,
    "BreakConstructorInitializers": "BeforeColon",
    "BreakConstructorInitializersBeforeComma": false,
    "BreakInheritanceList": "BeforeColon",
    "BreakStringLiterals": false,
    "ColumnLimit": 0,
    "CommentPragmas": "",
    "CompactNamespaces": false,
    "ConstructorInitializerAllOnOneLineOrOnePerLine": false,
    "ConstructorInitializerIndentWidth": 2,
    "ContinuationIndentWidth": 2,
    "Cpp11BracedListStyle": false,
    "DeriveLineEnding": true,
    "DerivePointerAlignment": true,
    "DisableFormat": false,
    "EmptyLineAfterAccessModifier": "Leave",
    "EmptyLineBeforeAccessModifier": "Leave",
    "ExperimentalAutoDetectBinPacking": false,
    "FixNamespaceComments": false,
    "ForEachMacros": [
        "foreach",
        "Q_FOREACH",
        "BOOST_FOREACH"
    ],
    "IfMacros": [
        "KJ_IF_MAYBE"
    ],
    "IncludeBlocks": "Preserve",
    "IncludeCategories": [
        {
            "Regex": "^\"(llvm|llvm-c|clang|clang-c)/",
            "Priority": 2,
            "SortPriority": 0,
            "CaseSensitive": false
        },
        {
            "Regex": "^(<|\"(gtest|gmock|isl|json)/)",
            "Priority": 3,
            "SortPriority": 0,
            "CaseSensitive": false
        },
        {
            "Regex": ".*",
            "Priority": 1,
            "SortPriority": 0,
            "CaseSensitive": false
        }
    ],
    "IncludeIsMainRegex": "",
    "IncludeIsMainSourceRegex": "",
    "IndentAccessModifiers": false,
    "IndentCaseBlocks": true,
    "IndentCaseLabels": true,
    "IndentExternBlock": "Indent",
    "IndentGotoLabels": false,
    "IndentPPDirectives": "None",
    "IndentRequires": true,
    "IndentWidth": 2,
    "IndentWrappedFunctionNames": false,
    "InsertTrailingCommas": "None",
    "JavaScriptQuotes": "Leave",
    "JavaScriptWrapImports": true,
    "KeepEmptyLinesAtTheStartOfBlocks": true,
    "LambdaBodyIndentation": "Signature",
    "Language": "Cpp",
    "MacroBlockBegin": "",
    "MacroBlockEnd": "",
    "MaxEmptyLinesToKeep": 100000,
    "NamespaceIndentation": "None",
    "ObjCBinPackProtocolList": "Auto",
    "ObjCBlockIndentWidth": 2,
    "ObjCBreakBeforeNestedBlockParam": true,
    "ObjCSpaceAfterProperty": false,
    "ObjCSpaceBeforeProtocolList": true,
    "PPIndentWidth": -1,
    "PackConstructorInitializers": "BinPack",
    "PenaltyBreakAssignment": 1,
    "PenaltyBreakBeforeFirstCallParameter": 1,
    "PenaltyBreakComment": 1,
    "PenaltyBreakFirstLessLess": 1,
    "PenaltyBreakOpenParenthesis": 1,
    "PenaltyBreakString": 1,
    "PenaltyBreakTemplateDeclaration": 1,
    "PenaltyExcessCharacter": 1,
    "PenaltyIndentedWhitespace": 1,
    "PenaltyReturnTypeOnItsOwnLine": 1,
    "PointerAlignment": "Right",
    "QualifierAlignment": "Leave",
    "ReferenceAlignment": "Pointer",
    "ReflowComments": false,
    "RemoveBracesLLVM": false,
    "SeparateDefinitionBlocks": "Leave",
    "ShortNamespaceLines": 0,
    "SortIncludes": "Never",
    "SortJavaStaticImport": "Before",
    "SortUsingDeclarations": false,
    "SpaceAfterCStyleCast": false,
    "SpaceAfterLogicalNot": false,
    "SpaceAfterTemplateKeyword": false,
    "SpaceAroundPointerQualifiers": "Default",
    "SpaceBeforeAssignmentOperators": true,
    "SpaceBeforeCaseColon": false,
    "SpaceBeforeCpp11BracedList": false,
    "SpaceBeforeCtorInitializerColon": true,
    "SpaceBeforeInheritanceColon": true,
    "SpaceBeforeParens": "ControlStatements",
    "SpaceBeforeParensOptions": {
        "AfterControlStatements": true,
        "AfterForeachMacros": true,
        "AfterFunctionDefinitionName": false,
        "AfterFunctionDeclarationName": false,
        "AfterIfMacros": true,
        "AfterOverloadedOperator": false,
        "BeforeNonEmptyParentheses": false
    },
    "SpaceBeforeRangeBasedForLoopColon": true,
    "SpaceBeforeSquareBrackets": false,
    "SpaceInEmptyBlock": false,
    "SpaceInEmptyParentheses": false,
    "SpacesBeforeTrailingComments": 2,
    "SpacesInAngles": "Leave",
    "SpacesInCStyleCastParentheses": false,
    "SpacesInConditionalStatement": false,
    "SpacesInContainerLiterals": false,
    "SpacesInLineCommentPrefix": {
        "Minimum": 0,
        "Maximum": -1
    },
    "SpacesInParentheses": false,
    "SpacesInSquareBrackets": false,
    "Standard": "Auto",
    "StatementAttributeLikeMacros": [
        "Q_EMIT"
    ],
    "StatementMacros": [
        "Q_UNUSED",
        "QT_REQUIRE_VERSION"
    ],
    "TabWidth": 2,
    "UseCRLF": false,
    "UseTab": "Never",
    "WhitespaceSensitiveMacros": [
        "STRINGIZE",
        "PP_STRINGIZE",
        "BOOST_PP_STRINGIZE",
        "NS_SWIFT_NAME",
        "CF_SWIFT_NAME"
    ]
}
```

</p>
</details>

**Comment by per1234** [comment]
Thanks so much @kittaakos. That approach is ideal for facilitating the maintenance of the formatter configuration.

I have now updated the PR according to your suggestions:

https://github.com/arduino/arduino-ide/compare/0719836e98c37634817a8b63cb004a8f6c933bd6..fe411f591d757e8a5b22076b3bf057974127feb1

File: arduino-ide-extension/src/node/clang-formatter.ts [hunk 1]
```diff
@@ -147,189 +147,10 @@ function styleJson({
   UseTab,
 }: ClangFormatOptions): Record<string, unknown> {
   // Source: https://github.com/arduino/tooling-project-assets/tree/main/other/clang-format-configuration
+  const defaultConfig = require('./default-formatter-config.json'); // 1. require the JSON
```
Comment (kittaakos) [inline arduino-ide-extension/src/node/clang-formatter.ts:150]: This line should be changed to, it's my fault. I found it only when verifying the behavior:

```ts
const defaultConfig = require('../../src/node/default-formatter-config.json');
```

Explanation:

We write TS code that will be transpiled to JS code. IDE2 does not run TS code, only JS code. The sources (TS files, JSON, .etc) are in `src`, the transpiled JS code is in `lib`.

The TS sources and the JSON config are here:
```
arduino-ide-extension/src/node/clang-formatter.ts
arduino-ide-extension/src/node/default-formatter-config.json
```

The TS code will be JS, and the module location will be these:
```
arduino-ide-extension/lib/node/clang-formatter.d.ts
arduino-ide-extension/lib/node/clang-formatter.d.ts.map
arduino-ide-extension/lib/node/clang-formatter.js
```

Notice the `src` and `lib` differences in the path. JSON files will remain in `src` at runtime, JS files will be in `lib` folder. To be able to `require` the JSON file from the transpiled JS, the `require` path must be changed from `./default-formatter-config.json` to `../../src/node/default-formatter-config.json`. If this is not clear, please let me know.

Also, please remove the `// 1. require the JSON` and the `// 2. override the default values with the user-defined ones` comments. I added them for explanation purposes.

Thank you!

Comment (per1234) [reply arduino-ide-extension/src/node/clang-formatter.ts]: >This line should be changed

Thanks. I have done it here: https://github.com/arduino/arduino-ide/pull/1324/commits/becb9013da4f915aec3e138bb10a14903698b901

>JSON files will remain in src at runtime, JS files will be in lib folder.

I saw that when I went to test the build (I think I somehow ended up still old source code when I ran it from source).

I was going down the path mentioned here:

https://stackoverflow.com/a/53629381/7059512

(adding something like `"src/node/*.json"` to the `include` field of `tsconfig.json`)

>please remove the // 1. require the JSON and the // 2. override the default values with the user-defined ones comments.

OK, done here: https://github.com/arduino/arduino-ide/pull/1324/commits/aff356078cfa066215b445cc925d918876a6d5e2

**Review by kittaakos — APPROVED** [review]
It's working very nicely. Thank you!

----

https://user-images.githubusercontent.com/1405703/184913662-83387391-82c3-4e4b-b182-aea98bf1f557.mp4

**Pull request merged**